### ISSUE TITLE
Do not assume public artifacts have prefix public/

### ISSUE THREAD

**Issue by petemoore** [taskcluster/taskgraph#407]
Searching taskgraph codebase, there are several pieces of code that check whether an artifact name has prefix `public/`. Unfortunately this is not the correct way to determine whether an API call to fetch the artifact requires scopes.

The anonymous role determines the set of scopes that all calls are granted. Typically this includes `queue:get-artifact:public/*` but that is not a requirement, and in locked down taskcluster deployments this scope may not be included in the anonymous role. Furthermore, the anonymous role may include other scopes with prefix `queue:get-artifact:` meaning that not only `public/*` artifacts are public.

The bottom line is: the correct way to check whether an artifact is public is to see if an unauthorized request satisfies the scope  `queue:get-artifact:<artifact-name>` e.g. by calling [authorize]( https://github.com/taskcluster/taskcluster/blob/8e083bd94d44b1b73dae604abb4bffb97145fdcd/libraries/api/src/middleware/auth.js#L227-L233) (node.js) or e.g. [scopes.Satisfies](https://github.com/taskcluster/taskcluster/blob/main/internal/scopes/scopes.go#L61) (go). I suspect there is a python library hiding somewhere too.

**Comment by ahal** [comment]
Thanks for the report, makes sense.

I do worry that making a network request for each artifact we need to check will be expensive. I wonder if in some cases an educated guess like what we're doing here is good enough. Or maybe there's a way to query the prefixes the anonymous role uses once and go with that?

**Comment by petemoore** [comment]
> Thanks for the report, makes sense.
> 
> I do worry that making a network request for each artifact we need to check will be expensive. I wonder if in some cases an educated guess like what we're doing here is good enough. Or maybe there's a way to query the prefixes the anonymous role uses once and go with that?

Good point! In python, you can cache the results of the anonymous role, and the scopeMatch routine I think doesn't make any external network request.
